cql3/restrictions/statement_restrictions.cc

```diff
@@ -1043,6 +1047,381 @@ expression make_conjunction(const expression& a, const expression& b) {
     return conjunction{children};
 }
 
+namespace {
+
+using cql3::selection::selection;
+
+/// Serialized values for all types of cells, plus selection (to find a column's index) and options (for
+/// subscript term's value).
+struct row_data {
+    const std::vector<bytes>& partition_key;
+    const std::vector<bytes>& clustering_key;
+    const std::vector<bytes_opt>& other_columns;
+    const selection& sel;
+    const query_options& options;
+};
+
+/// Returns col's value from the fetched data.
+bytes_opt get_value(const column_value& col, row_data data) {
+    auto cdef = col.col;
+    if (col.sub) {
+        auto col_type = static_pointer_cast<const collection_type_impl>(cdef->type);
+        if (!col_type->is_map()) {
+            throw exceptions::invalid_request_exception("subscripting non-map column");
+        }
+        const auto deserialized = cdef->type->deserialize(*data.other_columns[data.sel.index_of(*cdef)]);
+        const auto& data_map = value_cast<map_type_impl::native_type>(deserialized);
+        const auto key = col.sub->bind_and_get(data.options);
+        auto&& key_type = col_type->name_comparator();
+        const auto found = with_linearized(*key, [&] (bytes_view key_bv) {
+            return std::find_if(data_map.cbegin(), data_map.cend(), [&] (auto&& element) {
+                return key_type->compare(element.first.serialize_nonnull(), key_bv) == 0;
+            });
+        });
+        return found == data_map.cend() ? bytes_opt() : bytes_opt(found->second.serialize_nonnull());
+    } else {
+        switch (cdef->kind) {
+        case column_kind::partition_key:
+            return data.partition_key[cdef->id];
+        case column_kind::clustering_key:
+            return data.clustering_key[cdef->id];
+        case column_kind::static_column:
+        case column_kind::regular_column:
+            return data.other_columns[data.sel.index_of(*cdef)];
+        default:
+            throw exceptions::unsupported_operation_exception("Unknown column kind");
+        }
+    }
+}
+
+/// The comparator type for cv.
+data_type comparator(const column_value& cv) {
+    return cv.sub ?
+            static_pointer_cast<const collection_type_impl>(cv.col->type)->value_comparator() : cv.col->type;
```

Comment:
Should the cast not be to a reference? Isn't this creating a copy (and a slicing copy, too)

Reply:
No, it's fine, it's not a static_cast but a static_pointer_cast. My mistake.